### DISCUSSION TITLE
How to rebuild a page instance and put it back to the pdf with the results of page.get_text("json")?

### DISCUSSION THREAD

**Discussion by lidh15** [pymupdf/PyMuPDF#1461] [Q&A] [answered]
the case is simply from I need to remove some blocks in some pages of the pdf file. It's not difficult to get the json string and manipulate the blocks in the dict, but after that I want to recreate the pdf file, is there any API available?

**Comment by JorjMcKie** [comment, marked as answer]
Don't take the JSON string, take the "dict" format instead. JSON is derived from it.

You cannot write back to the same page - instead make a new page in a new PDF and write (a subset of) the contents of the dictioanry to it.
There is an example set of [scripts](https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/font-replacement) for font replacement, where you can probably learn a lot from.

**Reply by lidh15** [reply]
So another question is that how to find which block is on the top? I was able to remove a block, but I need to know which block is on the top thus gonna be shown and then I can remove the others.

**Comment by lidh15** [comment]
And I have another curious topic, how to transfer the image block we get into a PIL image instance like return by PIL.Image.open?

**Reply by JorjMcKie** [reply]
Image blocks within `get_text("dict")` have the key `"image"` the **_value_** of which is the binary image content (so may be large!).
> you also get a range of image meta information, including the matrix that was used to transform the original image rectangle into its bbox on the page.

Make a Pixmap from it `pix = fitz.Pixmap(block["image"])`, which you can save to a Pillow iamge via `pix.pil_save(... pillow parameters ...)`, or via `pix.pil_tobytes(...)`.

A potential issue here is that transparent images (with an alpha channel) are not properly handled. This would need special treatment - let me know when this happens.

**Reply by lidh15** [reply]
this is nice, in fact my question is about the mapping between fitz colorspace and pil image mode.
I searched related issues and saw this issue #322 , I guess that means we have 'gray', 'rgb' and 'cmyk' in fitz and when we `pil_tobytes` an image, the result pil image can only be one of these three mode 'I', 'RGB', 'CMYK' respectively.
the problem is:
firstly, as you mentioned, if there is a transparent png in the pdf, it might be 'rgba' which is not one of the three fitz color above, what should the mode be when I call `Image.frombytes`?
secondly, how do I know the fitz color as well as pil mode from the block? I see we have "color" keyword in the block, but the value is only an interger, how to map it into a pil mode string?

**Reply by JorjMcKie** [reply]
The `pil_save()` checks the pixmap's colorspace and generates Pillow modes automatically from it.
This code is in Python, so you could simply take and modify it if you want the Pillow.Image and not the `tobytes()` result. Or look at the tutorial in the documentation to see how this is done.

To find out whether the image has a mask, things are a bit more complicated. As I wrote, MuPDF does not reflect this in their `TextPage` object unfortunately. But you can do this:
```python
>>> # the dict of an image block
>>> pprint(page.get_text("dict")["blocks"][0].keys())
dict_keys(['number', 'type', 'bbox', 'width', 'height', 'ext', 'colorspace', 'xres', 'yres', 'bpc', 'transform', 'size', 'image'])
>>> # the dict of an image info: identify image from previous by its "number"
>>> pprint(page.get_image_info(xrefs=True)[0].keys())
dict_keys(['number', 'bbox', 'transform', 'width', 'height', 'colorspace', 'cs-name', 'xres', 'yres', 'bpc', 'size', 'digest', 'xref'])
>>> # look up "xref" in this list (1014). If second item is > 0, we have a transparent image
>>> pprint(page.get_images())
[(1014, 1015, 1945, 1004, 8, 'DeviceRGB', '', 'Im1', 'DCTDecode')]
>>> # now build complete transparent pixmap
>>> pix1014 = fitz.Pixmap(doc, 1014)  # base image
>>> pix1015 = fitz.Pixmap(doc, 1015)  # image mask
>>> pix = fitz.Pixmap(pix1014, pix1015)  # pixmap with alpha
>>> pix1014 = pix1015 = None  # free up storage
```

**Reply by lidh15** [reply]
Here is my problem, 
1. when I tried to use `doc.get_image_list(0)` to get the images from the sample file, it returned an empty list.
2. when I tried to modify the font, the code seemed to modify the pictures at the same time and broke the file totally. To replicate this, just pass the condition in line 240, repl-font.py as following. I expect it to remove all the text but left the other things unchanged. but the saved pdf is not as expected.

```
    def remove_font(fontrefs, lines):
        """This inline function removes references to fonts in a /Contents stream.

        Args:
            fontrefs: a list of bytes objects looking like b"/fontref ".
            lines: a list of the lines of the /Contents.
        Returns:
            (bool, lines), where the bool is True if we have changed any of
            the lines.
        """
        changed = False
        count = len(lines)
        for ref in fontrefs:
            found = False  # switch: processing our font
            for i in range(count):
                if lines[i] == b"ET":  # end text object
                    found = False  # no longer in found mode
                    continue
                if lines[i].endswith(b" Tf"):  # font invoker command
                    if True:
                    # if lines[i].startswith(ref):  # our font?
                        found = True  # switch on
                        lines[i] = b""  # remove line
                        changed = True  # tell we have changed
                        continue  # next line
                    else:  # else not our font
                        found = False  # switch off
                        continue  # next line
                if found == True and (
                    lines[i].endswith(
                        (
                            b"TJ",
                            b"Tj",
                            b"TL",
                            b"Tc",
                            b"Td",
                            b"Tm",
                            b"T*",
                            b"Ts",
                            b"Tw",
                            b"Tz",
                            b"'",
                            b'"',
                        )
                    )
                ):  # write command for our font?
                    lines[i] = b""  # remove it
                    changed = True  # tell we have changed
                    continue
        return changed, lines
```
here is the link of the sample file:
https://drive.google.com/file/d/12r-R7SyLP1HifWHN1f60wTc1ZeVdXN6Q/view?usp=sharing

**Reply by JorjMcKie** [reply]
> when I tried to use doc.get_image_list(0) to get the images from the sample file, it returned an empty list.

But it still does have images! Only if you use `page.get_image_info(...)` you will get all images the page displays - including the so-called "inline" images as in this case.

**Reply by JorjMcKie** [reply]
Else admittedly, font replacement does not work in all cases. Here, the non-embedded fonts with non-ASCII names are a problem, that I will look after some time.

**Reply by JorjMcKie** [reply]
Here is a snippet that removes the inline images. Once this is done, font replacement will work ok:
```python
import fitz

doc = fitz.open("test.pdf")
page = doc[0]
page.clean_contents()
cont_xref = page.get_contents()[0]
cont = bytearray(doc.xref_stream(cont_xref))
while True:
    p1 = cont.find(b"BI")
    if p1 < 0:  # no (more) inline image found
        break
    p2 = cont.find(b"EI", p1)
    cont[p1 : p2 + 2] = b""  # remove inline image
doc.update_stream(cont_xref, cont)
doc.ez_save("test-no-images.pdf")
```
A less hacky way would of course be removing the inline images by using redaction annotations!
Walk through `page.get_image_info()` and create a redaction annotation for every image bbox. Then do `page.apply_redactions(images=fitz.PDF_REDACTION_IMAGE_REMOVE)`.

**Reply by JorjMcKie** [reply]
If you modify the JSON output of font replacement like this, the resulting PDF will look good ... and exhibit text that was hidden under those inline images 😎.
```json
[
  {
    "oldfont": [
      "\u008b\u00cc\u00e5,Bold",
      "\ufffd\ufffd\ufffd,Bold"
    ],
    "newfont": "cjk",  # the built-in universal font supporting Chinese
    "info": "Not embedded!"
  },
  {
    "oldfont": [
      "\u00bf\u00ac\u00cc\u00e5_GB2312,Bold",
      "\ufffd\ufffd\ufffd\ufffd_GB2312,Bold"
    ],
    "newfont": "cjk",
    "info": "Not embedded!"
  },
  {
    "oldfont": [
      "\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd\ufffd,Bold",
      "\u00bb\u00aa\u0087\u00bf\u00ac\u00cc\u00e5,Bold"
    ],
    "newfont": "cjk",
    "info": "Not embedded!"
  }
]
```

**Reply by lidh15** [reply]
Thank you, I followed this and it worked, but I can only open the modified pdf with Chrome, when I tried to open it with adobe pdf reader it said that there were errors in the page and adobe acrobat cannot render the page correctly. The reader somehow opened it but the text written by "fitz.text_writer" and the images inserted by "page.insert_image" were not seen. Is there any clues for the reason why this happened?

**Reply by JorjMcKie** [reply]
Try to do it again with the non-hacky method (using redactions). It is "official" and not that error-prone.
My Adobe has no complaints, so you probably made some kind of error.
Otherwise I know that Adobe is very picky.

**Reply by lidh15** [reply]
Is that related to the arguments of doc.save()? 
I keep using the settings from the repl-font.py, and I'm wondering how it makes differences when I have different value for "deflate" and "garbage"? What arguments combination can provide the smallest file size?

**Reply by JorjMcKie** [reply]
> What arguments combination can provide the smallest file size?

There is a `save` alias called "easy save", `ez_save()` with some different default settings `garbage=3, deflate=True`. Which is a good balance between size reduction and save speed. Use garbage=4 to achieve the top compression (with MuPDF).

**Comment by JorjMcKie** [comment]
> how to find which block is on the top?

When MuPDF builds its TextPage object (which I wrap with the same-named Python object), it applies some heuristics when it creates its block-line-span-character hierarchy.
This process leaves the original sequence of the underlying PDF painting instructions intact - at least roughly.
> Some minor deviations may happen depending on the `flags` parameter, which may cause space characters to be generated or general whitespace be replaced by one or more spaces.

So, in the generated dictionary `get_text("dict", ...)`, the sequence of the text pieces should reflect the sequence of the corresponding original painting commands. If text bboxes overlap, one might conclude that a **_later_** one overlaps an **_earlier_** one.

I also have `page.get_bboxlog()` which is a list of rectangles flagged with the rectangle "type" (text, drawing, image). Their sequence indeed is the painting sequence.
This is a more complete picture - i.e. not only regarding text, but all sort of objects potentially overlapping each other.
